db/queries/base.py

```diff
@@ -24,6 +25,7 @@ def __init__(
         self.engine = engine
         self.transformations = transformations
         self.name = name
+        self.metadata = metadata
```

Comment:
This can result in stale metadata; makes `DBQuery` stateful. If you create a `DBQuery` with metadata returned by `get_cached_metadata`, then call `reset_reflection`, and then do something with that `DBQuery` instance from earlier, it will still use the old metadata that was supposed to be invalidated by `reset_reflection`, but wasn't because `DBQuery` is effectively caching it. Could you make the `initial_relation` property a method like `get_initial_relation(metadata)`, and make sure that `DBQuery` operations always use up-to-date metadata?

Reply:
I am okay with passing `metadata` as a parameter but I am curious to know why `metadata` shouldn't be part of `DBQuery`. `DBQuery` is stateful, it has access to `engine` instance, shouldn't it be the same case with `metadata`?

I am suggesting that `DBQuery` should not be reused or we should be rethinking how the metadata caching actually works(should this be a singleton).

Reply:
Whether `engine` should be static in a `DBQuery` is debatable. Here it's sort of treated as a reference to a database, so in that sense it is static. Notice that we don't actually mutate engines and we don't expect them to be mutated, but we do mutate metadata. If you fetch a cached metadata instance, it's only meant to be valid at that point in time, so it shouldn't be cached again.

Not sure what you mean when you say that a DBQuery should be a singleton, or do you mean `MetaData`?

I don't think there's a problem with reusing `DBQuery`, unless it's stateful.

Reply:
> Not sure what you mean when you say that metadata should be a singleton.

We don't have a need for multiple `Metadata` objects in our case.

Usually, when dealing with caches, the cache object would be a singleton and the same instance would be accessed by the codebase, and if the cache has to be cleared, we would be calling a method of the object responsible for clearing the cache(in this case [sqlalchemy.schema.MetaData.clear()](https://docs.sqlalchemy.org/en/14/core/metadata.html#sqlalchemy.schema.MetaData)) this avoids passing metadata to each method

Reply:
Can we move this refactor to a different issue? Changing `initial_relation` to a function would trigger various refactors as other properties make use of this property

Reply:
If you're proposing a cache on the `db` namespace, I'd say that any kind of state is inappropriate in `db`. In `mathesar`, we still need multiple concurrent instances of metadata, because we can't yet reuse metadata everywhere. And, hopefully, we'll never have to, since that would be a lot of work.

If by singleton you mean having only a single instance of `MetaData` at any given time, that's not possible currently (in `mathesar` namespace) or not desirable (in `db` namespace`).

> [sqlalchemy.schema.MetaData.clear()](https://docs.sqlalchemy.org/en/14/core/metadata.html#sqlalchemy.schema.MetaData)

That's an interesting idea. Currently the cached metadata is either a `MetaData` or `None`, so you'd have to go from that to it always being `MetaData`.

> Can we move this refactor to a different issue?

I'm hesitant. I wouldn't want to run into this unexpected statefulness while working on something else. A bug involving this wouldn't be easy to find.

If you decide to postpone fixing this, I won't oppose it, but please create an issue to track this. Also, a warning in a code comment next to the `metadata` attribute might be helpful.